File: src/swift/Sources/Signalbird/SignalbirdClient.swift
import Foundation

/// Beş seviye. Fazlası eklenmez: kanal ayarını anlaşılır tutar.
public enum SignalbirdLevel: String, Sendable {
    case debug, info, warn, error, critical
}

public struct SignalbirdConfig {
    /// Sunucu anahtarı (`sb_secret_live_…`).
    ///
    /// Bu anahtar GİZLİDİR: iOS uygulamasına gömülmemelidir — uygulama paketi
    /// çözülebilir ve anahtar herkesin eline geçer. Mobil taraf sohbeti ve push
    /// kaydını `SignalbirdApp` ile (açık anahtarla) yapar; Telsiz istemcisi
    /// sunucu tarafı Swift servisleri (Vapor, Hummingbird) içindir.
    public let domainKey: String
    public let baseURL: String
    public let source: String?
    public let timeout: TimeInterval
    public let throwOnError: Bool

    public init(
        domainKey: String,
        baseURL: String = signalbirdDefaultBaseURL,
        source: String? = nil,
        timeout: TimeInterval = 5,
        throwOnError: Bool = false
    ) {
        self.domainKey = domainKey
        self.baseURL = baseURL
        self.source = source
        self.timeout = timeout
        self.throwOnError = throwOnError
    }
}

/// Telsiz (Radio) istemcisi — log ve olay yazar.
///
/// Sözleşme: docs/CONTRACT.md § 1–7
public struct SignalbirdClient {
    private let http: Transport
    private let source: String?

    public init(config: SignalbirdConfig) throws {
        guard !config.domainKey.isEmpty else {
            throw SignalbirdError(code: "NO_KEY", status: 0, message: "domainKey zorunlu")
        }

        // Açık anahtarın sunucuda kullanılması sessiz bir güvenlik hatasıdır.
        guard !config.domainKey.hasPrefix("sb_public_live_") else {
            throw SignalbirdError(
                code: "WRONG_KEY_TYPE",
                status: 0,
                message: "sunucu istemcisi sb_secret_live_… ister"
            )
        }

        self.source = config.source
        self.http = Transport(
            baseURL: config.baseURL,
            timeout: config.timeout,
            throwOnError: config.throwOnError
        ) { ["Authorization": "Bearer \(config.domainKey)"] }
    }

    /// Seviye verilmezse kanalın kendi varsayılanı geçerlidir.
    @discardableResult
    public func log(
        key: String,
        message: String,
        level: SignalbirdLevel? = nil,
        context: [String: Any]? = nil
    ) async throws -> SbResult {
        try await http.request("POST", "/v1/radio/log", body: [
            "key": key,
            "message": message,
            "level": level?.rawValue as Any,
            "context": context as Any,
            "source": source as Any,
        ])
    }

    @discardableResult
    public func debug(_ key: String, _ message: String, _ context: [String: Any]? = nil) async throws -> SbResult {
        try await log(key: key, message: message, level: .debug, context: context)
    }

    @discardableResult
    public func info(_ key: String, _ message: String, _ context: [String: Any]? = nil) async throws -> SbResult {
        try await log(key: key, message: message, level: .info, context: context)
    }

    @discardableResult
    public func warn(_ key: String, _ message: String, _ context: [String: Any]? = nil) async throws -> SbResult {
        try await log(key: key, message: message, level: .warn, context: context)
    }

    @discardableResult
    public func error(_ key: String, _ message: String, _ context: [String: Any]? = nil) async throws -> SbResult {
        try await log(key: key, message: message, level: .error, context: context)
    }

    @discardableResult
    public func critical(_ key: String, _ message: String, _ context: [String: Any]? = nil) async throws -> SbResult {
        try await log(key: key, message: message, level: .critical, context: context)
    }

    /// En fazla 100 kayıt, satır satır sonuç. Başarısız satırlar YENİDEN DENENMEZ.
    @discardableResult
    public func batch(_ events: [[String: Any]]) async throws -> SbResult {
        let rows = events.prefix(100).map { event -> [String: Any] in
            var row = event
            if row["source"] == nil, let source { row["source"] = source }

            return row
        }

        return try await http.request("POST", "/v1/radio/log/batch", body: ["events": Array(rows)])
    }
}

File: bin/main/io/signalbird/sdk/SignalbirdClient.kt
package io.signalbird.sdk

/** Beş seviye. Fazlası eklenmez: kanal ayarını anlaşılır tutar. */
enum class SignalbirdLevel(val value: String) {
    DEBUG("debug"),
    INFO("info"),
    WARN("warn"),
    ERROR("error"),
    CRITICAL("critical"),
}

data class SignalbirdConfig(
    /**
     * Sunucu anahtarı (`sbr_live_…`).
     *
     * Bu anahtar GİZLİDİR ve Android uygulamasına gömülmemelidir — APK
     * çözülebilir. Mobil taraf sohbeti ve push kaydını [SignalbirdApp] ile
     * (açık anahtarla) yapar; bu istemci sunucu tarafı Kotlin servisleri
     * (Ktor, Spring) içindir.
     */
    val apiKey: String,
    val baseUrl: String = SIGNALBIRD_DEFAULT_BASE_URL,
    /** Her olaya eklenen köken adı (sunucu ya da servis adı). */
    val source: String? = null,
    /** Varsayılan 5 sn — bir log çağrısı isteği bekletmemeli. */
    val timeoutMs: Int = 5_000,
    /** Varsayılan kapalı: telsiz erişilemezse ödeme akışı çökmemeli. */
    val throwOnError: Boolean = false,
)

/**
 * Telsiz (Radio) istemcisi — log ve olay yazar.
 *
 * Sözleşme: docs/CONTRACT.md § 1–7
 */
class SignalbirdClient(private val config: SignalbirdConfig) {
    private val http: Transport

    init {
        require(config.apiKey.isNotEmpty()) { "Signalbird: apiKey zorunlu." }

        // Açık anahtarın sunucuda kullanılması sessiz bir güvenlik hatasıdır:
        // çalışır görünür, sonra kanal kısıtına takılır. Baştan söylüyoruz.
        require(!config.apiKey.startsWith("sbr_pub_")) {
            "Signalbird: sunucu istemcisine tarayıcı anahtarı verildi. sbr_live_… kullanın."
        }

        http = Transport(config.baseUrl, config.timeoutMs, config.throwOnError) {
            mapOf("Authorization" to "Bearer ${config.apiKey}")
        }
    }

    /** Seviye verilmezse kanalın kendi varsayılanı geçerlidir. */
    suspend fun log(
        channel: String,
        message: String,
        level: SignalbirdLevel? = null,
        context: Map<String, Any?>? = null,
    ): SbResult = http.request(
        "POST",
        "/v1/radio/log",
        mapOf(
            "channel" to channel,
            "message" to message,
            "level" to level?.value,
            "context" to context,
            "source" to config.source,
        ),
    )

    suspend fun debug(channel: String, message: String, context: Map<String, Any?>? = null): SbResult =
        log(channel, message, SignalbirdLevel.DEBUG, context)

    suspend fun info(channel: String, message: String, context: Map<String, Any?>? = null): SbResult =
        log(channel, message, SignalbirdLevel.INFO, context)

    suspend fun warn(channel: String, message: String, context: Map<String, Any?>? = null): SbResult =
        log(channel, message, SignalbirdLevel.WARN, context)

    suspend fun error(channel: String, message: String, context: Map<String, Any?>? = null): SbResult =
        log(channel, message, SignalbirdLevel.ERROR, context)

    suspend fun critical(channel: String, message: String, context: Map<String, Any?>? = null): SbResult =
        log(channel, message, SignalbirdLevel.CRITICAL, context)

    /**
     * En fazla 100 kayıt, satır satır sonuç.
     *
     * Kısmi başarı normaldir (kota tam ortada dolabilir). Başarısız satırlar
     * YENİDEN DENENMEZ: aynı logu iki kez yazmak da bir maliyettir.
     */
    suspend fun batch(events: List<Map<String, Any?>>): SbResult {
        val rows = events.take(100).map { event ->
            if (event["source"] == null && config.source != null) {
                event + ("source" to config.source)
            } else {
                event
            }
        }

        return http.request("POST", "/v1/radio/log/batch", mapOf("events" to rows))
    }
}
